File: leetcode/leetcode/Solutions/TwoPointers/ContainerWithMostWater.swift
//
//  ContainerWithMostWater.swift
//  leetcode
//
//  Created by Kai Wang on 8/5/22.
//  Copyright © 2022 Pirate. All rights reserved.
//

import Foundation

class ContainerWithMostWater {
    func maxArea(_ height: [Int]) -> Int {
        var maxArea = 0
        var left = 0
        var right = height.count - 1
        while left < right {
            let area = (right - left) * min(height[left], height[right])
            maxArea = max(area, maxArea)
            if height[left] < height[right] {
                left += 1
            } else {
                right -= 1
            }
        }
        return maxArea
    }

    func test() {
        let case1 = [1,8,6,2,5,4,8,3,7]
        let case2 = [1,1]
        let cases = [case1, case2]
        for input in cases {
            print(maxArea(input))
        }
    }
}

File: leetcode/Python/TwoPointers/ContainerWithMostWater.py
from typing import List


class Solution:
    def maxArea(self, heights: List[int]) -> int:
        left, right = 0, len(heights) - 1
        max_volume = 0
        while left != right:
            volume = min(heights[left], heights[right]) * (right - left)
            max_volume = max(volume, max_volume)
            if heights[left] <= heights[right]:
                left += 1
            else:
                right -= 1
        return max_volume
